library/std/src/sys/pal/unix/sync/condvar.rs

```diff
@@ -84,6 +81,68 @@ impl Condvar {
+// extension that measures timeouts based on the monotonic clock and is thus
+// resilient against wall-clock changes.
+#[cfg(target_vendor = "apple")]
+impl Condvar {
+    /// # Safety
+    /// * `init` must have been called on this instance.
+    /// * `mutex` must be locked by the current thread.
+    /// * This condition variable may only be used with the same mutex.
+    pub unsafe fn wait_timeout(&self, mutex: Pin<&Mutex>, dur: Duration) -> bool {
+        let mutex = mutex.raw();
+
+        // The macOS implementation of `pthread_cond_timedwait` internally
+        // converts the timeout passed to `pthread_cond_timedwait_relative_np`
+        // to nanoseconds. Unfortunately, the "psynch" variant of condvars does
+        // not guard against overflow during the conversion[^1], which means
+        // that `pthread_cond_timedwait_relative_np` will return `ETIMEDOUT`
+        // much earlier than expected if the relative timeout is longer than
+        // `u64::MAX` nanoseconds.
+        //
+        // This can be observed even on newer platforms (by setting the environment
+        // variable PTHREAD_MUTEX_USE_ULOCK to a value other than "1") by calling e.g.
+        // ```
+        // condvar.wait_timeout(..., Duration::from_secs(u64::MAX.div_ceil(1_000_000_000));
+        // ```
+        // (see #37440, especially
+        // https://github.com/rust-lang/rust/issues/37440#issuecomment-3285958326).
+        //
+        // To work around this issue, always clamp the timeout to u64::MAX nanoseconds,
+        // even if the "ulock" variant is used (which does guard against overflow).
+        //
+        // [^1]: https://github.com/apple-oss-distributions/libpthread/blob/1ebf56b3a702df53213c2996e5e128a535d2577e/kern/kern_synch.c#L1269
+        const MAX_DURATION: Duration = Duration::from_nanos(u64::MAX);
+
+        let (dur, clamped) = if dur <= MAX_DURATION { (dur, false) } else { (MAX_DURATION, true) };
+
+        let timeout = libc::timespec {
+            // This cannot overflow because of the clamping above.
+            tv_sec: dur.as_secs() as i64,
+            tv_nsec: dur.subsec_nanos() as i64,
+        };
+
+        // This function has been available since macOS version 10.4
+        // and iOS version 2.0 (i.e. the first versions with any
+        // pthread API).
+        // FIXME: move this definition to libc.
+        unsafe extern "C" {
+            unsafe fn pthread_cond_timedwait_relative_np(
+                cond: *mut libc::pthread_cond_t,
+                lock: *mut libc::pthread_mutex_t,
+                timeout: *const libc::timespec,
+            ) -> libc::c_int;
+        }
+
+        let r = unsafe { pthread_cond_timedwait_relative_np(self.raw(), mutex, &timeout) };
+        assert!(r == libc::ETIMEDOUT || r == 0);
+        // Report clamping as a spurious wakeup. Who knows, maybe some
+        // interstellar space probe will rely on this ;-).
+        r == 0 || clamped
```

Comment:
This is actually an interesting API question... is `wait_timeout` allowed to return "timeout" for a spurious wakeup?

The other codepath, which clamps the timeout on cygwin, still assumes that to be allowed, but this codepath here goes out of its way to avoid that.

Reply:
The documentation of `wait_timeout` says that (emphasis mine)
> The returned [WaitTimeoutResult](https://doc.rust-lang.org/nightly/std/sync/struct.WaitTimeoutResult.html) value indicates if the timeout is *known* to have elapsed.

So I'd say that spurious wakeups may not return a `WaitTimeoutResult` that indicates timeout. But then again, who knows what the system might do – the newer futex-like macOS implementation of `pthread_cond_timedwait` for instance properly clamps the nanoseconds, but will not perform this check. I guess it isn't too much of an issue, given that even the clamped timeout will only actually occur after 584 (macOS) and 1000 (Cygwin) years.